### PR TITLE
Make Jitsi WCAG 2.1 compliant

### PR DESCRIPTION
**Pull request by skolmer**
# WCAG 2.1 AA Compliance / Accessibility

This PR is the followup to https://github.com/jitsi/jitsi-meet/pull/8423 @saghul @damencho 

We managed to make jitsi-meet (web) fully WCAG 2.1 AA compliant. This has been tested and approved by two independent entities. 
One feedback I want to share with the community came from a person who relies on a screen-reader for his daily work. He was very excited that he was finally able to participate in video conferences at his workplace.

We would really appreciate if you could give this review some priority because it looks like there is currently a lot of work being done on UI elements and we already had a hard time keeping up with the master branch while doing our final code review 🙂

This is a rather large PR (sorry 😉) because we had to modify almost every element of jitsi-meet web. But don't be afraid there are a lot of similar changes like aria attributes, color contrasts or keypress handler. 
I will go into more detail regarding PTT and global keyboard shortcuts below to give you more insights what we did and why we did it. This should make your life easier reviewing all these changes. If there are still things we missed to explain please let us know in the comments.

## Features

- Full keyboard access to all UI elements
- Better contrast for improved readability/visibility on some elements
- Descriptive aria attributes and labels for screen readers
- Highlighting of focused elements
- Unified wording / More descriptive labels
- A setting to disable keyboard shortcuts

## Fixes

- No interference between input elements like buttons and PTT or other global keyboard shortcuts
- Hide Mute Audio/Video moderation features from toolbox overflow menu if user is not a moderator
- Re-integrated video mute button to toolbox overflow menu
- Auto-Hide of toolbox will now recognize focused elements and stay open until unfocused
- Auto-Hide of toolbox will now recognize when a user is hovering the remote-video-control menu
- Fixed all places where `eslint-disable-next-line react/jsx-no-bind` was used to avoid unnecessary rerender
- The <kbd>SPACE</kbd> key will now only unmute while pressed and you were muted before. The old behavior was to also activate mute on <kbd>SPACE</kbd> similar to the <kbd>M</kbd> key. If you take push-to-talk literally this was a wrong behavior and <kbd>M</kbd> should be the only shortcut to mute while <kbd>SPACE</kbd> will allow you to talk while holding <kbd>SPACE</kbd> when muted.
- Minor UI fixes like z-index, padding issues, missing headlines in dialogs, responsiveness or overflowing elements

## Push-To-Talk (PTT) / Keyboard Shortcuts

We decided on an approach that allowed us to keep global shortcuts (like PTT) the jitsi community is used to while adding space bar support to UI elements like buttons and menus.
This took some extra work because PTT and other shortcuts like numbers can interfere with the expected experience of screen-reader users and keyboard users in general.

1. We allow people to disable keyboard shortcuts entirely if features like PTT or number shortcuts are causing too much confusion when using a screen-reader for example. This is a new checkbox in the settings dialog. Keyboard shortcuts stay active by default though to make this a non-breaking change.
2. We prioritized the keyboard handling of keyboard-focused elements (like input, button, checkbox, ...) over global keyboard shortcuts. Which means when you intentionally focus a button using the tab key the function of the space bar will be to trigger the button. You are also now able to remove focus from an element by pressing ESC. This allows keyboard users to have the native web experience they are used to while also having great features like PTT.

## Credits

Big thanks to my team who did most of the hard work @agnesboldt @ahmadkadri @MaikHannemann @nurjinjafar @phk-nord 

Dataport would like to thank the Jitsi Open Source Community for a great product with this active contribution from its Phoenix Open Source Initiative – for further information please visit https://www.phoenix-werkstatt.de or https://www.dataport.de

### CODE REVIEW

**Comment by skolmer** [comment]
> 
> 
> Jenkins, test this please.

We expected some failing tests regarding changed toolbox button labels. If I remember correctly they are used to select the DOM elements in some tests. Can we move to test-id attributes instead to avoid issues in the future or will this cause to many issues because of failing tests on other branches?

**Comment by saghul** [comment]
Amazing work! Please bear with us as we review such a large amount of code. I suppose the answer is "no", but I'll try anyway: any chance this can be split into smaller units of work for easier review?

**Comment by skolmer** [comment]
> 
> 
> Amazing work! Please bear with us as we review such a large amount of code. I suppose the answer is "no", but I'll try anyway: any chance this can be split into smaller units of work for easier review?

Unfortunately no, these changes all kind of depended on each other so we squashed them all together.

**Comment by damencho** [comment]
I will give it a try today, to fix the tests.

**Comment by damencho** [comment]
I gave it a quick test and I see that PTT does not work, other shortcuts seems to work (at least video mute).

**Comment by skolmer** [comment]
> 
> 
> I gave it a quick test and I see that PTT does not work, other shortcuts seems to work (at least video mute).

It will not mute you on first press of the spacebar (previous behavior) instead it will only unmute you when you hold space while you are muted. This was kind of confusing behavior before because the expectation for push to talk was that holding SPACE will unmute you while you are muted. <- was this what happened in your case?

**Comment by damencho** [comment]
Yes, exactly. That's it ...

**Comment by damencho** [comment]
@skolmer Hey, the author needs to be the same for the PRs to work together, and the branch names of course. Can you open a PR in jitsi-meet-torture, please? 
Like this one, the branch name, but needs to be from nordeck https://github.com/jitsi/jitsi-meet-torture/pull/372. Thanks.

**Comment by skolmer** [comment]
> 
> 
> Yes, exactly. That's it ...

I have added a better description to "Fixes" in the PR description to make this more obvious. I'm aware this is a change that can cause confusion If you are used to the old behavior and tried to better explain our interpretation how PTT should work. Let us know if you can follow our argumentation or if this needs some more discussion.

**Comment by damencho** [comment]
Oh and first conflicts came in, sorry for that. 

So one other thing I see, so the tests are failing around the room password and setting password on pre join screen, I will take a look at the tests ... but one change in behaviour, room key/password field, when setting - `Enter` was specially handled to set the password/key where now it closes the Dialog and actually does not set a password to the room. And since we moved the security button away from the toolbar this makes it a real concern as people may think that password is set after hitting Enter, where they just dismissed the Dialog. Can you look at that? ... I will try checking the tests tomorrow.

**Comment by damencho** [comment]
> Hide Mute Audio/Video moderation features from toolbox overflow menu if user is not a moderator
This was already the case, or I'm missing something?

**Comment by skolmer** [comment]
> but one change in behaviour, room key/password field, when setting - `Enter` was specially handled to set the password/key where now it closes the Dialog and actually does not set a password to the room.

This definitely sounds like a regression. I will take a look tomorrow.

**Comment by skolmer** [comment]
> 
> 
> > Hide Mute Audio/Video moderation features from toolbox overflow menu if user is not a moderator
> 
> This was already the case, or I'm missing something?

This was an interesting bug I noticed while testing you can still see this on alpha.jitsi.net. There was a check like `visible && isModerator` so when the `visible` prop of the button was not set (which was the case) it never evaluated the `isModerator` flag and defaulted to the default visbility of the base component which was `true`. So the simple fix here was to give `visible` a default of `true` in the inheriting component too.

**Comment by damencho** [comment]
> > but one change in behaviour, room key/password field, when setting - `Enter` was specially handled to set the password/key where now it closes the Dialog and actually does not set a password to the room.
> 
> This definitely sounds like a regression. I will take a look tomorrow.

Thanks, yeah the passwords tests are failing because of that. https://github.com/jitsi/jitsi-meet-torture/blob/7537ee65a7d43248e9c3a42a2e353be3e8de129d/src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java#L88

**Comment by damencho** [comment]
Yep, all other tests are failng because of https://github.com/jitsi/jitsi-meet-torture/blob/7537ee65a7d43248e9c3a42a2e353be3e8de129d/src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java#L88, just checked and Lobby.

**Comment by skolmer** [comment]
> 
> 
> Yep, all other tests are failng because of https://github.com/jitsi/jitsi-meet-torture/blob/7537ee65a7d43248e9c3a42a2e353be3e8de129d/src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java#L88, just checked and Lobby.

Thanks for testing and taking the time to explain, sounds like we will be able to provide a fix tomorrow.

**Comment by skolmer** [comment]
> 
> 
> Yep, all other tests are failng because of https://github.com/jitsi/jitsi-meet-torture/blob/7537ee65a7d43248e9c3a42a2e353be3e8de129d/src/test/java/org/jitsi/meet/test/pageobjects/web/SecurityDialog.java#L88, just checked and Lobby.

This is now fixed

**Comment by damencho** [comment]
Jenkins, test this please.

**Comment by muscat1** [comment]
Holy moly, this is a formidable amount of work! Freaking awesome, if I may say.

I've done some manual testing of the PR and I want to drop my findings here in hopes that you and everyone else reviewing will agree / disagree with them. I'll then also do a run through the code.

Here they go:

1. The copy link section in the prejoin screen looks different, I noticed you reused the `<CopyMeetingLinkSection />` component but I'm not sure that's what we want, was there a reason for this?

Screenshot 2021-04-05 at 10 43 29

2. The active speaker border colour was changed, is this intended for a better visibility? To me it looks a little bit out of place, but feel free to disagree.

Screenshot 2021-04-05 at 10 43 40

3. You've added the 'Chat' back in the chat header, but that was purposefully removed not too long ago due to a UX request. Does it influence the chat accessibility in any way?

Screenshot 2021-04-05 at 10 38 14

4. Also due to a UX request, we've removed the autofocus on the chat input but you seem to have added it back, I can imagine that it has some accessibility implications, but I wanted to confirm whether that is actually needed or not.

5. We also want to keep the input field in focus after sending a message in chat, but that is currently broken and an error is thrown (can be seen in the inspector).

6. Another pretty big one is that trying to open chat on mobile browser effectively crashes the app and hangs up the meeting. Please keep mobile browser in mind when making changes as it is quite high on our priority list to keep things smooth on both desktop and mobile web.

7. Still on the chat business, the emojis are accessible via keyboard but I couldn't seem to send one via keyboard.

Screenshot 2021-04-05 at 10 38 43

8. Lastly, when accessing chat without a nickname, you are prompted to enter one via the `<DisplayNameForm />` component, inside which the "Enter chat room" button is not accessible via keyboard.

Screenshot 2021-04-05 at 10 38 03

**Comment by skolmer** [comment]
@muscat1 Thank you for you detailed testing and feedback! 
Before I comment on your findings just one quick question. Because of all the issues you had with chat it looks to me like your local test setup was using an old version of `react-textarea-autosize`. Can you run npm install again and let me know if these issues persist? Thanks!

**Comment by muscat1** [comment]
Yikes, you are more than correct, @skolmer, thanks! I've edited my comment above and struck through the points that are no longer applicable.

As a side note, apparently, by using tab to select and then _spacebar_ , you can actually use the keyboard to input an emoji from the panel, but I'm not sure if that is obvious to the user 🤷🏻‍♂️

**Comment by vp8x8** [comment]
One small comment for the prejoin page:
Screenshot 2021-04-05 at 15 22 20


If The "Don't show this screen again" text is hovered the cursor is not changed to hand(pointer). If the area around is, that's ok but the text is problematic (maybe because a `label` is used).

**Comment by skolmer** [comment]
@muscat1 Below my feedback to your points:

1. The current behavior is this: 
image
So it is hardly identifiable as a button nor does it give the user any information about what the button is doing. We added a label and gave it the same style as other buttons (blue for more contrast). When adding keypress handlers and everything we realized there is already another component that does exactly the same. So to reduce complexity we used the same component here.

2. We had to find a way to have this border always visible even if the user is presenting in front of a blue wall or showing a bluish picture. The challenge was to find a box shadow that works on any background and with every color the current frame of the users video shows. We decided on a box shadow that consists of a black inner shadow and a white outer one to always have sufficient contrast.

3. Sections of a website should have headers. This is a good general rule - not only for screen reader users - to give a user more context about what's currently in front of the user. The chat is a good example for that. If you open the chat to the side it shows you a message box where you can type in your message but it doesn't really explain what this panel is used for. This might be obvious if the panel is not open by default and you clicked on a button "Show Chat" before. But even than context should not depend on memorizing the actions you did before, it should be given by the site structure. I'm not aware of the discussion that took place before removing the header but if it was only for "design" reasons I would strongly recommend to keep it.

4. For accessibility reasons it is a good practice to move focus to a dialog after it is opened. For chat there might be a better solution or even some concerns we should discuss first.

7. Yep, I'm also not 100% happy with the current solution but it passed the accessibility tests. So it made things usable. There are ways to improve, for example using real emojis in the textbox too with labels describing them to screen reader users. But regarding space and tab for selection these are known controls for keyboard users but I think we can add Enter support too. The rules when to use space or enter or both are mainly given by the role and the default behavior of standard html elements. Regarding Tab control it would be better to have the emoji selector after the text-field for a more natural tab flow - not having to tab back. I would be happy for any suggestions how to improve here but would keep it for another PR.

8. Looks like this came in with one of the latest updates from master and we missed it. I just added a keypress handler and other accessibility attributes. https://github.com/jitsi/jitsi-meet/pull/8921/commits/53d86eef5b4bee3bd1cc569a5d6c487a21b6c0b8

File: react/features/recording/components/LiveStream/web/StreamKeyPicker.js
Comment (skolmer) [reply react/features/recording/components/LiveStream/web/StreamKeyPicker.js:114]: This was done to get rid of `eslint-disable-next-line react/jsx-no-bind`. Before this change it was creating new arrow functions inline on each render which can cause unnecessary rerenders because react has no way to detect that the new arrow function passed to the event handler prop is basically doing the same as the old one. We added a static event handler function that now gets the streamId from the element that is given in the event args of the keypress or click event.

File: css/_avatar.scss
Comment (skolmer) [reply css/_avatar.scss:29]: Good point. I will check if it can be safely removed and is not set by another css rule.

File: modules/keyboardshortcut/keyboardshortcut.js
Comment (skolmer) [reply modules/keyboardshortcut/keyboardshortcut.js]: Good catch. It has been optimized here https://github.com/jitsi/jitsi-meet/pull/8921/commits/6630a2e8cfe270d2e8da1f663f6cac4a6e629c6e

File: react/features/toolbox/actions.web.js
Comment (skolmer) [reply react/features/toolbox/actions.web.js:92]: What solution were you thinking about? It is intended that this checks the currently rendered DOM in realtime.

**Comment by skolmer** [comment]
> 
> 
> One small comment for the prejoin page:
> Screenshot 2021-04-05 at 15 22 20
> 
> If The "Don't show this screen again" text is hovered the cursor is not changed to hand(pointer). If the area around is, that's ok but the text is problematic (maybe because a `label` is used).

Good catch 👍 This is now fixed too. https://github.com/jitsi/jitsi-meet/pull/8921/commits/9335a0d00999b0d70c90aad57151a3fd6d1a9eb9

File: react/features/chat/components/web/ChatInput.js
Comment (skolmer) [reply react/features/chat/components/web/ChatInput.js:133]: In this case you need to run `npm install` because you still have an old version of `react-textarea-autosize` installed.

File: css/modals/settings/_settings.scss
Comment (skolmer) [reply css/modals/settings/_settings.scss:27]: These are css vars used by the atlaskit component. 
image

File: css/_avatar.scss
Comment (skolmer) [reply css/_avatar.scss]: Fixed

File: css/_labels.scss
Comment (skolmer) [reply css/_labels.scss]: Fixed

File: css/_subject.scss
Comment (skolmer) [reply css/_subject.scss]: Fixed

File: css/modals/invite/_info.scss
Comment (skolmer) [reply css/modals/invite/_info.scss]: Fixed

**Comment by saghul** [comment]
Quick heads up @skolmer : we have a relatively large PR close to landing which creates a new panel where some functionality is moved. It will likely create some conflicts when landed and may need some accessibility review.

**Comment by skolmer** [comment]
> 
> 
> Quick heads up @skolmer : we have a relatively large PR close to landing which creates a new panel where some functionality is moved. It will likely create some conflicts when landed and may need some accessibility review.

Thanks for the heads up.

This brings up an interesting question. How are you planning to ensure things stay accessible after this PR got merged? Should new contribution rules be defined that at least keyboard control and aria attributes should be implemented when creating new UI elements? There are also eslint plugins that can give some security that accessibility doesn't break on new contributions (https://reactjs.org/docs/accessibility.html#eslint-plugin-jsx-a11y) but in the end it might be best to implement some internal review guidelines about what needs to be considered regarding accessibility when new UI elements are developed or old UI is restructured.

**Comment by saghul** [comment]
> This brings up an interesting question. How are you planning to ensure things stay accessible after this PR got merged?

❤️  Thanks for pointing it out. TBH I was hoping you'd help us figure that out :-)



> Should new contribution rules be defined that at least keyboard control and aria attributes should be implemented when creating new UI elements?

We can certainly put that in the CONTRIBUTING.md file.

> There are also eslint plugins that can give some security that accessibility doesn't break on new contributions (https://reactjs.org/docs/accessibility.html#eslint-plugin-jsx-a11y) but in the end it might be best to implement some internal review guidelines about what needs to be considered regarding accessibility when new UI elements are developed or old UI is restructured.

Humans are forgetful and can make mistakes, but machines (usually) don't. So I'd say let's use tooling to our advantage, where possible.

File: react/features/toolbox/actions.web.js
Comment (vp8x8) [reply react/features/toolbox/actions.web.js:92]: Something like keeping focus state of the elements of the selector (e.g. toolbox-content-items, filmstrip etc) in the store and use it here. But I don't actually know how easy/viable is to do it...

Ideally we would fully manage state in react. I don't think is a very good idea mixing DOM query with react state. WDYT?

Comment (skolmer) [reply react/features/toolbox/actions.web.js:92]: In this case I think it would be less efficient to manage this in state because there can are a lot of element's involved that need to be tracked here. Basically all elements in Toolbox and Overflow Menus, Sidebar and more. So I believe using a DOM selector to just get elements with `:focus` or `:focus-within` is more efficient even if it doesn't feel right for a react app. I would argue Focus is a state of the rendered DOM and not necessarily one of the React Component.

Comment (vp8x8) [reply react/features/toolbox/actions.web.js:92]: Fair enough.

**Comment by saghul** [comment]
General note: please avoid making UI changes. We'd need to sync with design, think of the right way forward, come back to you, wait for it to be updated, rinse and repeat. This would take too long and put this PR at risk since it's so large and involved. So let's first focus on non-UI behavior (for the most part) and once that lands we can discuss each individual UI component separately. This will help get the work merged quicker.

>     1. So it is hardly identifiable as a button nor does it give the user any information about what the button is doing. We added a label and gave it the same style as other buttons (blue for more contrast). When adding keypress handlers and everything we realized there is already another component that does exactly the same. So to reduce complexity we used the same component here.

Please drop this change and let's discuss changing that UI separately.


>     2\. We had to find a way to have this border always visible even if the user is presenting in front of a blue wall or showing a bluish picture. The challenge was to find a box shadow that works on any background and with every color the current frame of the users video shows. We decided on a box shadow that consists of a black inner shadow and a white outer one to always have sufficient contrast.

Ditto.



>     3\. Sections of a website should have headers. This is a good general rule - not only for screen reader users - to give a user more context about what's currently in front of the user. The chat is a good example for that. If you open the chat to the side it shows you a message box where you can type in your message but it doesn't really explain what this panel is used for. This might be obvious if the panel is not open by default and you clicked on a button "Show Chat" before. But even then context should not depend on memorizing the actions you did before, it should be given by the site structure. I'm not aware of the discussion that took place before removing the header but if it was only for "design" reasons I would strongly recommend to keep it.

Ditto.



>     4\. For accessibility reasons it is a good practice to move focus to a dialog after it is opened. For chat there might be a better solution or even some concerns we should discuss first.

The mobile usual UX is the opposite. We are discussing having different behavior depending on the platform, so please skip this in this PR.

>     5\. Yep, I'm also not 100% happy with the current solution but it passed the accessibility tests. So it made things usable. There are ways to improve, for example using real emojis in the textbox too with labels describing them to screen reader users. But regarding space and tab for selection these are known controls for keyboard users but I think we can add Enter support too. The rules when to use space or enter or both are mainly given by the role and the default behavior of standard html elements. Regarding Tab control it would be better to have the emoji selector after the text-field for a more natural tab flow - not having to tab back. I would be happy for any suggestions how to improve here but would keep it for another PR.

@muscat1 Do you have concerns about this still? I noticed you stroked your comment through.



>     6\. Looks like this came in with one of the latest updates from master and we missed it. I just added a keypress handler and other accessibility attributes. [53d86ee](https://github.com/jitsi/jitsi-meet/commit/53d86eef5b4bee3bd1cc569a5d6c487a21b6c0b8)

Cool!

Closing note: this needs a rebase too.

**Comment by saghul** [comment]
@skolmer FYI we landed that large change I told you about. Feel free to rebase and drop us a line when it's ready for review.

**Comment by saghul** [comment]
Jenkins add to whitelist.

File: react/features/virtual-background/components/VirtualBackgroundDialog.js
Comment (vp8x8) [inline react/features/virtual-background/components/VirtualBackgroundDialog.js:175]: These callbacks look very similar. I think a custom hook  could be created to avoid repetition.

File: react/features/participants-pane/components/MeetingParticipantItem.js
Comment (vp8x8) [inline react/features/participants-pane/components/MeetingParticipantItem.js:42]: I think there is the need for some params here. Maybe create a custom hook.
Check /jitsi-meet/react/features/base/i18n/functions.js.

**Comment by skolmer** [comment]
Quick heads up.. we are currently just waiting for your signal that the review is completed before we solve any more merge conflicts. I hope you understand that otherwise we risk to waste time in fixing merge conflicts in components that change over and over again. 
I think there are still some open questions you wanted to discuss with your design department so please let us know if we should make any changes to the styling. Also not sure about the mobile consideration - should we implement a check that it is a desktop browser before activating auto-focus in chat and are there other concerns regarding mobile?

Please let us know how we can proceed with this PR to get it merged without loosing part of the WCAG compliance requirements on the way. We will of course support the project in defining code or review guidelines to keep the project WCAG compliant in the future. Just let me know how you prefer to move forward with this.

**Comment by damencho** [comment]
We will let you know. Most of the team is off till Tuesday :) Thank you

**Comment by vp8x8** [comment]
@skolmer Please go ahead with the rebase. After looking over it the state of the UI is reasonable enough. 

As for mobile, having a check for desktop browser seems like a good idea. I don't see any other concerns at the moment.

Once the rebase is ready we'll take a final look over it (test it) and merge it.
Thanks for the patience, we're excited for this.

**Comment by skolmer** [comment]
> 
> 
> @skolmer Please go ahead with the rebase. After looking over it the state of the UI is reasonable enough.
> 
> As for mobile, having a check for desktop browser seems like a good idea. I don't see any other concerns at the moment.
> 
> Once the rebase is ready we'll take a final look over it (test it) and merge it.
> Thanks for the patience, we're excited for this.

Thanks, we just solved the latest merge conflicts.

**Comment by damencho** [comment]
@skolmer There is some strange error, not sure from where it is coming from though:
```
Run npm install
  npm install
  shell: /usr/bin/bash -e {0}
npm ERR! Unexpected end of JSON input while parsing near '...izHPDA="
npm ERR!     }
npm ERR!   }
npm ERR! }
npm ERR! '

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/runner/.npm/_logs/2021-06-09T16_16_30_637Z-debug.log
Error: Process completed with exit code 1.
```

**Comment by skolmer** [comment]
> 
> 
> @skolmer There is some strange error, not sure from where it is coming from though:
> 
> ```
> Run npm install
>   npm install
>   shell: /usr/bin/bash -e {0}
> npm ERR! Unexpected end of JSON input while parsing near '...izHPDA="
> npm ERR!     }
> npm ERR!   }
> npm ERR! }
> npm ERR! '
> 
> npm ERR! A complete log of this run can be found in:
> npm ERR!     /home/runner/.npm/_logs/2021-06-09T16_16_30_637Z-debug.log
> Error: Process completed with exit code 1.
> ```

Looks like we messed up the lockfile again. because of the newer npm version we are using we had to do a manual merge. 
Sorry for that, should be working now.

**Comment by damencho** [comment]
> Looks like we messed up the lockfile again. because of the newer npm version we are using we had to do a manual merge.
> Sorry for that, should be working now.

No worries ... I have no idea why the Github action is failing 🤷‍♂️ 
But there are other errors while executing the PR-tests: http://pr-reports.jitsi.org/jitsi-meet-PR-test/9462/build-logs/jitsi-meet-common.txt

**Comment by skolmer** [comment]
> But there are other errors while executing the PR-tests: http://pr-reports.jitsi.org/jitsi-meet-PR-test/9462/build-logs/jitsi-meet-common.txt

At least I remember now why we added this id prop to the device list... 🙈 This one is now fixed too.

The action error is strange. No clue what's wrong there looks like some internal error of the github workflow execution.

**Comment by skolmer** [comment]
The tests are now showing one error because we changed the aria-label attribute of the participant invite button to have the same value as the button-label itself. The best would be to change the element selector in the test from `[aria-label="Invite people"]` to `[aria-label="Invite Someone"]`

**Comment by damencho** [comment]
> The tests are now showing one error because we changed the aria-label attribute of the participant invite button to have the same value as the button-label itself. The best would be to change the element selector in the test from `[aria-label="Invite people"]` to `[aria-label="Invite Someone"]`

Can you change it actually in https://github.com/jitsi/jitsi-meet-torture/pull/373, thanks.

**Comment by damencho** [comment]
There will be one for the other PR and I will schedule and rerun for this one

**Comment by skolmer** [comment]
> 
> 
> There will be one for the other PR and I will schedule and rerun for this one
> Correction: the other PR is not whitelisted ... so we are testing only this, and is currently running.

I'm a little bit confused by the test that is now failing. 
It looks like it is testing the toolbox buttons and trying to click the invite button in the toolbox menu... but this button is only visible when 'invite' is in the TOOLBAR_BUTTONS array. Which is not the case. At least I can't find this config in the torture project. 
So maybe it was not testing the toolbar button it should have but the invite button in the sidebar instead, which is now renamed? We didn't change the toolbar invite button and it always had a different label (`Invite someone`) so looks like this test never tested the right button.

**Comment by damencho** [comment]
> > There will be one for the other PR and I will schedule and rerun for this one
> > Correction: the other PR is not whitelisted ... so we are testing only this, and is currently running.
> 
> I'm a little bit confused by the test that is now failing.
> It looks like it is testing the toolbox buttons and trying to click the invite button in the toolbox menu... but this button is only visible when 'invite' is in the TOOLBAR_BUTTONS array. Which is not the case. At least I can't find this config in the torture project.
> So maybe it was not testing the toolbar button it should have but the invite button in the sidebar instead, which is now renamed? We didn't change the toolbar invite button and it always had a different label (`Invite someone`) so looks like this test never tested the right button.

Hum I see it fails on org.jitsi.meet.test.pageobjects.web.InviteDialog.close(InviteDialog.java:102)` humm seems like it tries to click on a button away from the modal dialog expecting it to close .. and it doesn't 
org jitsi meet test InviteTest-web participant1

Not sure why ...

**Comment by damencho** [comment]
Is there a change around that dialog .... ?

**Comment by skolmer** [comment]
> 
> 
> Is there a change around that dialog .... ?

Not recently. But there was a change in the ModalHeader we had to merge in. https://github.com/jitsi/jitsi-meet/commit/c4677be87a26826120d559e5a592f5d23a0a70f8#diff-5dc31303b04eb9e5630274591a2dcbd1a4a6df33db756d64530cf7c0ec7bf893R86

I've looked in the test log and it says
> [11] TestFailure:
org.openqa.selenium.NoSuchElementException: no such element: Unable to locate element: {"method":"css selector","selector":"[aria-label="Invite people"]"}

This would indicate an issue locating the button that opens the dialog. In the stacktrace it also shows the InviteDialog.close though
> 	at org.openqa.selenium.remote.RemoteWebDriver.findElement(RemoteWebDriver.java:315)
	at org.jitsi.meet.test.pageobjects.web.Toolbar.getInviteButton(Toolbar.java:151)
	at org.jitsi.meet.test.pageobjects.web.InviteDialog.close(InviteDialog.java:102)
	at org.jitsi.meet.test.InviteTest.testInviteURLDisplays(InviteTest.java:69)

**Comment by damencho** [comment]
@skolmer can you update torture once again, please. For some reason I cannot push directly.
https://github.com/jitsi/jitsi-meet-torture/blob/00973955a6e44b7af6ae7d86f790a7989d9449bb/src/test/java/org/jitsi/meet/test/pageobjects/web/Toolbar.java#L41

**Comment by damencho** [comment]
I tested it locally and it seems to work after changing that.

**Comment by damencho** [comment]
Thanks, I started the tests again 🤞

**Comment by damencho** [comment]
Tadaaa green! @muscat1 and @vp8x8 your turn now. If you decide to squash and merge it, make sure you merge and the torture one. Thank you.

**Comment by vp8x8** [comment]
Jenkins please test this please.

**Comment by muscat1** [comment]
Took it for another spin and the conference behaves properly! 🚀

On a separate note, as Saul has stated waaay above, in the future, we should probably try to avoid such PRs as they are a nightmare both to review and to maintain. 

Waiting on the ok from @vp8x8 as well, but in the meantime, the CI seems to be failing for some reason 🤔

**Comment by skolmer** [comment]
> On a separate note, as Saul has stated waaay above, in the future, we should probably try to avoid such PRs as they are a nightmare both to review and to maintain.

I totally agree. I think it was the nature of this change that required to extend almost every component with keypress handler and aria tags. Next time we will try to do it component after component even if it means to have a 2-digit number of PRs... still better than maintaining THIS. 😊 Thanks for taking the time though

**Comment by muscat1** [comment]
Yeah, I think multiple small PRs would be much easier to treat in isolation.

Regarding the failing pipeline, it might have something to do with some discrepancy between package and package-lock. Could you possibly try to run a clean `npm i` with a lower version of node, 12, as well as npm 6 (We need to upgrade 🙄) as that might fix it.

**Comment by skolmer** [comment]
> 
> 
> Yeah, I think multiple small PRs would be much easier to treat in isolation.
> 
> Regarding the failing pipeline, it might have something to do with some discrepancy between package and package-lock. Could you possibly try to run a clean `npm i` with a lower version of node, 12, as well as npm 6 (We need to upgrade 🙄) as that might fix it.

Not today I'm afraid. But if you send me a fixed lockfile I can push it to the branch.

**Comment by vp8x8** [comment]
Test it and looks great. So when the pipeline issue is fixed we can merge. 🎊

**Comment by skolmer** [comment]
> 
> 
> Test it and looks great. So when the pipeline issue is fixed we can merge. 🎊

OK cool, CI seems to be running now 🎊

**Comment by damencho** [comment]
The tests were green yesterday when I left the message ... The only failing is the github action, but it is something on github side so, I'm not worried after the PR test passed.

**Comment by damencho** [comment]
@skolmer after merging this, some tool for detecting accessibility problems will be good to add so we keep the UI up-to-date, maybe you can help with that :)

**Comment by skolmer** [comment]
> 
> 
> @skolmer after merging this, some tool for detecting accessibility problems will be good to add so we keep the UI up-to-date, maybe you can help with that :)

My time in the next two weeks will be very limited but after that I will be happy to integrate some tooling and also make some suggestion how the rules for future feature PRs can look like. Tooling is limited so we will need a combination of tooling and a checklist of best practices a contributor/reviewer should follow.

**Comment by damencho** [comment]
👍

**Review by vp8x8 — APPROVED** [review]

**Review by muscat1 — APPROVED** [review]
Astounding work! Thanks for taking the time and having patience with the entire review process! 🚀🎊

**Comment by damencho** [comment]
Congrats @skolmer and Thank you!

Looking forward to implementing something automated + manual so we can fix and keep the code compliant. Thank you once again for the dedication and help.

**Pull request merged**
